src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/PoolingAsyncValueTaskMethodBuilderT.cs

```diff
@@ -0,0 +1,473 @@
+
+            /// <summary>Completes the box with a result.</summary>
+            /// <param name="result">The result.</param>
+            public void SetResult(TResult result) =>
+                _valueTaskSource.SetResult(result);
+
+            /// <summary>Completes the box with an error.</summary>
+            /// <param name="error">The exception.</param>
+            public void SetException(Exception error) =>
+                _valueTaskSource.SetException(error);
+
+            /// <summary>Gets the status of the box.</summary>
+            public ValueTaskSourceStatus GetStatus(short token) => _valueTaskSource.GetStatus(token);
+
+            /// <summary>Schedules the continuation action for this box.</summary>
+            public void OnCompleted(Action<object?> continuation, object? state, short token, ValueTaskSourceOnCompletedFlags flags) =>
+                _valueTaskSource.OnCompleted(continuation, state, token, flags);
+
+            /// <summary>Gets the current version number of the box.</summary>
+            public short Version => _valueTaskSource.Version;
+
+            /// <summary>Implemented by derived type.</summary>
+            TResult IValueTaskSource<TResult>.GetResult(short token) => throw NotImplemented.ByDesign;
+
+            /// <summary>Implemented by derived type.</summary>
+            void IValueTaskSource.GetResult(short token) => throw NotImplemented.ByDesign;
+        }
+
+        /// <summary>Type used as a singleton to indicate synchronous success for an async method.</summary>
+        private sealed class SyncSuccessSentinelStateMachineBox : StateMachineBox
+        {
+            public SyncSuccessSentinelStateMachineBox() => SetResult(default!);
+        }
+
+        /// <summary>Provides a strongly-typed box object based on the specific state machine type in use.</summary>
+        private sealed class StateMachineBox<TStateMachine> :
+            StateMachineBox,
+            IValueTaskSource<TResult>, IValueTaskSource, IAsyncStateMachineBox, IThreadPoolWorkItem
+            where TStateMachine : IAsyncStateMachine
+        {
+            /// <summary>Delegate used to invoke on an ExecutionContext when passed an instance of this box type.</summary>
+            private static readonly ContextCallback s_callback = ExecutionContextCallback;
+            /// <summary>Thread-local cache of boxes. This currently only ever stores one.</summary>
+            [ThreadStatic]
+            private static StateMachineBox<TStateMachine>? t_tlsCache;
+            /// <summary>Lock used to protected the shared cache of boxes. 1 == held, 0 == not held.</summary>
+            /// <remarks>The code that uses this assumes a runtime without thread aborts.</remarks>
+            private static int s_cacheLock;
+            /// <summary>Singly-linked list cache of boxes.</summary>
+            private static StateMachineBox<TStateMachine>? s_cache;
```

Comment:
@davidfowl, you were asking about pooling and whether it's one pool for all async methods or one pool per async method.  It's the latter: the pool is part of this `StateMachineBox<TStateMachine>` class, so it's unique per `TStateMachine`, which ends up being the type (a struct in release builds) the compiler generates for the async method.